server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```diff
@@ -446,8 +445,8 @@ static void parseObjectOrNested(ParseContext context, ObjectMapper mapper) throw
 
         innerParseObject(context, mapper, parser, currentFieldName, token);
         // restore the enable path flag
-        if (nested.isNested()) {
-            nested(context, nested);
+        if (mapper instanceof NestedObjectMapper) {
```

Comment:
All the `instanceof`s worry me it'll be very hard to look at `NestedObjectMapper` and figure out how it's used. I think we're doing future-us a disservice with them. If they are temporary just to break the type hierarchy that's cool, but maybe we should keep `isNested` and have it return `false` from `ObjectMapper` and `true` from `NestedObjectMapper`, at least in the short term? At least that way I can look at `ObjectMapper` and know "oh, there is this weird overrideable method that always returns `true`, I bet we're doing a funny thing with it".

Reply:
Yeah, they are a bit janky.  They should all go pretty soon, as my plan is to add a NestedLookup object built from mappings that *only* contains NestedObjectMapper and so we won't actually need to do any of this checking, but for now I will replace with an `isNested` method.

Reply:
That'd make my heart happy!

As a bonus, when zap the `isNested` method in a follow up we can be sure that you've cleaned all the sites that used to need `instanceof`. That'd be lovely.